consensus/enclave/impl/src/constant_time_token_map.rs

```diff
@@ -0,0 +1,331 @@
+//! This module defines a helper object for tracking and accumulating fees
+//! in the enclave, while remaining constant time with respect to token ids.
+//! This is used both when looking up the fee for a transaction, and when
+//! accumulating all the fees when forming a block.
+//!
+//! Note, in the future it may be more efficient to use data-structures / code
+//! from the mc-oblivious repo, which is optimized for x86-64.
+
+use alloc::vec::Vec;
+use core::{
+    iter::{FromIterator, IntoIterator},
+    ops::Add,
+};
+use mc_transaction_core::TokenId;
+use subtle::{Choice, ConditionallySelectable, ConstantTimeEq};
+
+// A CtTokenMap is a structure mapping token ids to values.
+// (For example, fees and minimum fees.)
+//
+// This structure supports get and set operations in constant time, by scanning
+// across the list of pairs.
+//
+// Keys are set upon initialization, and cannot be added after initialization.
+#[derive(Debug, Clone)]
+pub struct CtTokenMap<T: ConditionallySelectable + Default> {
+    storage: Vec<(TokenId, T)>,
+}
+
+impl<T: ConditionallySelectable + Default> CtTokenMap<T> {
+    /// Get the value associated to a token id in constant time.
+    ///
+    /// If the token id is not found, returns None.
+    ///
+    /// Note that we do not return CtOption, it is not a goal to be
+    /// constant-time with respect to whether the token is in the map, only
+    /// that we are constant-time with respect to which token id it is that
+    /// is in the map.
+    pub fn get(&self, token_id: &TokenId) -> Option<T> {
+        let mut success = Choice::from(0);
+        let mut result = T::default();
+        for (key, val) in self.storage.iter() {
+            let found = key.ct_eq(token_id);
+            result.conditional_assign(val, found);
+            success.conditional_assign(&found, found);
```

Comment:
This is subtle yet eloquent, I like it.

2 questions to elicit discussion:
1. What's the approx length of the constant time interval
2. So we have to do a max O(n) search every time?

How many lookup operations like this would we expect in normal operations of FOG or consensing? Like how often would you think these get/set ops would be triggered

Reply:
> What's the approx length of the constant time interval

If you're asking like, how much stuff needs to be constant time, I think the answer is like, if a user submits a transaction we have to make sure that any function that takes that token id does not reveal information about the token id via the access pattern. So mainly we're concerned with, when we have to look up something connected to the token id in a map, then we have to think about that and try to make that constant time.

The constant time window for this unction is just the for loop basically

> So we have to do a max O(n) search every time?

In this version, yeah. If we stuck a fancier oblivious data structure, then we wouldn't have to.
However, in the near future we only expect to have a handful of tokens. So it's `O(n)` but `n` is pretty small.
Even in a future where there are a lot of tokens, we may decide that not all of them can be used to pay fees,
and so they don't count into this `O(n)`.

> How many lookup operations like this would we expect in normal operations of FOG or consensing? Like how often would you think these get/set ops would be triggered

I think when we check if a `Tx` is well-formed, we look up the minimum fee for its token. So we would have one get operation for that. During `form_block`, I think we check all the minimum fees again. Then we do the fee accumulation thing, so that's like `O(n)` per transaction. Then we sort, so that's like `O(n^2)`.

There are low-level things we could do to make this faster, if we used the `aligned-cmov` crate instead of `subtle` that would be faster. But that entails bringing in the `mc-oblivious` repo into consensus enclave. Maybe fine but needs discussion.

I think for probably a few hundred tokens these overheads aren't going to be noticeable, that's my guess anyways.

Reply:
I'd love to archive this conversation in somewhere visible. It's absolutely on the money that this is the correct KISS solution. This is just great information for future engineers who might have to modify it on the happy day we reach a large amount of token ids & traffic.

Reply:
I think, at some point this stuff becomes tribal knowledge. Glad you are in the tribe :)

We can maybe do a journal club or something